File: T3/gramatica.java
import java.util.Scanner;

class gramatica{
    public static String[] name_m = {"cacador","cachorro","rio","mar","tempo","sino","vento","martelo","rosto","tambor"};
    public static String[] name_f= {"menina","noticia","floresta","porta","vida","mae","cidade"};
    public static String[] name_m_p = {"lobos","tambores"};
    public static String[] name_f_p = {"meninas","lagrimas"};
    public static String[] verb = {"corre","correu","bateu"};
    public static String[] verb_p  = {"corriam","bateram","correram"};
    public static String[] prep = {"para","com"};
    public static String[] contraction_f = {"na","pela","Na","Pela"};
    public static String[] contraction_m  = {"no","pelo","No","Pelo"};
    public static String[] contraction_m_p  = {"nos","pelos","Nos","Pelos"};
    public static String[] contraction_f_p  = {"nas","pelas","Nas","Pelas"};
    public static String[] determiner_m  = {"o","O"};
    public static String[] determiner_f  = {"a","A"};
    public static String[] determiner_m_p  = {"os","Os"};
    public static String[] determiner_f_p  = {"as","As"};

    public static boolean isName_m(String s){
	for(int i=0;i<name_m.length;i++)
	    if(s.equals(name_m[i]))
		return true;
	return false;
    }
    
    public static boolean isName_f(String s){
	for(int i=0;i<name_f.length;i++)
	    if(s.equals(name_f[i]))
		return true;
	return false;
    }
    
    public static boolean isName_m_p(String s){
	for(int i=0;i<name_m_p.length;i++)
	    if(s.equals(name_m_p[i]))
		return true;
	return false;
    }
    
    public static boolean isName_f_p(String s){
	for(int i=0;i<name_f_p.length;i++)
	    if(s.equals(name_f_p[i]))
		return true;
	return false;
    }

    public static boolean isVerb(String s){
	for(int i=0;i<verb.length;i++)
	    if(s.equals(verb[i]))
		return true;
	return false;
    }

    public static boolean isVerb_p(String s){
	for(int i=0;i<verb_p.length;i++)
	    if(s.equals(verb_p[i]))
		return true;
	return false;
    }

    public static boolean isPrep(String s){
	for(int i=0;i<prep.length;i++)
	    if(s.equals(prep[i]))
		return true;
	return false;
    }

    public static boolean isContraction_m(String s){
	for(int i=0;i<contraction_m.length;i++)
	    if(s.equals(contraction_m[i]))
		return true;
	return false;
    }

       public static boolean isContraction_f(String s){
	for(int i=0;i<contraction_f.length;i++)
	    if(s.equals(contraction_f[i]))
		return true;
	return false;
    }

       public static boolean isContraction_m_p(String s){
	for(int i=0;i<contraction_m_p.length;i++)
	    if(s.equals(contraction_m_p[i]))
		return true;
	return false;
    }

       public static boolean isContraction_f_p(String s){
	for(int i=0;i<contraction_f_p.length;i++)
	    if(s.equals(contraction_f_p[i]))
		return true;
	return false;
    }
    
    public static boolean isDeterminer_m(String s){
	for(int i=0;i<determiner_m.length;i++)
	    if(s.equals(determiner_m[i]))
		return true;
	return false;
    }

    public static boolean isDeterminer_f(String s){
	for(int i=0;i<determiner_f.length;i++)
	    if(s.equals(determiner_f[i]))
		return true;
	return false;
    }

      public static boolean isDeterminer_m_p(String s){
	for(int i=0;i<determiner_m_p.length;i++)
	    if(s.equals(determiner_m_p[i]))
		return true;
	return false;
    }

    public static boolean isDeterminer_f_p(String s){
	for(int i=0;i<determiner_f_p.length;i++)
	    if(s.equals(determiner_f_p[i]))
		return true;
	return false;
    }

    public static String isSentence(String[] s){
	String tree ="sentence(";
	int i=0;
	String singularity="";
	if(s.length<3 || s.length>6)
	    return "false";
	if(i<s.length && isDeterminer_m(s[i])){
	    singularity = "s";
	    if((i+1)<s.length && !isName_m(s[i+1]))
		return "false";
	}
	if(i<s.length && isDeterminer_f(s[i])){
	    singularity = "s";
	    if((i+1)<s.length && !isName_f(s[i+1]))
		return "false";
	}
	if(i<s.length && isDeterminer_m_p(s[i])){
	    singularity = "p";
	    if((i+1)<s.length && !isName_m_p(s[i+1]))
		return "false";
	}
	if(i<s.length && isDeterminer_f_p(s[i])){
	    singularity = "p";
	    if((i+1)<s.length && !isName_f_p(s[i+1]))
		return "false";
	}
	if(!singularity.equals(""))
	    tree = tree + "noun_phrase(determiner("+s[i]+"), name("+s[i+1]+")), ";
	else{
	    if(i<s.length && isContraction_m(s[i])){
		singularity = "s";
		if((i+1)<s.length && !isName_m(s[i+1]))
		    return "false";
	    }
	    if(i<s.length && isContraction_f(s[i])){
		singularity = "s";
		if((i+1)<s.length && !isName_f(s[i+1]))
		    return "false";
	    }
	
	    if(i<s.length && isContraction_m_p(s[i])){
		singularity = "p";
		if((i+1)<s.length && !isName_m_p(s[i+1]))
		    return "false";
	    }

	    if(i<s.length && isContraction_f_p(s[i])){
		singularity = "p";
		if((i+1)<s.length && !isName_f_p(s[i+1]))
		    return "false";
	    }
	    if(singularity.equals(""))
		return "false";
	    else
		tree = tree + "noun_phrase(contraction("+s[i]+"), name("+s[i+1]+")), ";
	}   
	i = 2;
	if(i<s.length && singularity.equals("s") && !isVerb(s[i]))
	    return "false";
	if(i<s.length && singularity.equals("p") && !isVerb_p(s[i]))
	    return "false";
	if(s.length==3){
	    tree = tree + "verbal_phrase(verb("+s[i]+")))";
	    return tree;
	}
	tree = tree + "verbal_phrase(verb("+s[i]+"), ";
	i = 3;
	if(i<s.length && isContraction_m(s[i])){
	    if((i+1)<s.length && isName_m(s[i+1])){
		tree = tree + "noun_phrase(contraction("+s[i]+"), name("+s[i+1]+"))))";
		return tree;
	    }
	    else
		return "false";
	}
	if(i<s.length && isContraction_f(s[i])){
	    if((i+1)<s.length && isName_f(s[i+1])){
		tree = tree + "noun_phrase(contraction("+s[i]+"), name("+s[i+1]+"))))";
		return tree;
	    }
	    else
		return "false";
	}
	
	if(i<s.length && isContraction_m_p(s[i])){
	    if((i+1)<s.length && isName_m_p(s[i+1])){
		tree = tree + "noun_phrase(contraction("+s[i]+"), name("+s[i+1]+"))))";
		return tree;
	    }
	    else
		return "false";
	}

	if(i<s.length && isContraction_f_p(s[i])){
	    if((i+1)<s.length && isName_f_p(s[i+1])){
		tree = tree + "noun_phrase(contraction("+s[i]+"), name("+s[i+1]+"))))";
		return tree;
	    }
	    else
		return "false";
	}

	if(i<s.length && isPrep(s[i])){
	    if((i+1)<s.length && (i+2)<s.length && isDeterminer_m(s[i+1]) && (!isName_m(s[i+2])))
		return "false";
	    if((i+1)<s.length && (i+2)<s.length && isDeterminer_m(s[i+1]) && isName_m(s[i+2])){
		tree = tree + "prep_phrase(preposition("+s[i]+"), noun_phrase(determiner("+s[i+1]+"),name("+s[i+2]+")))))";
		return tree;
	    }
	    if((i+1)<s.length && (i+2)<s.length && isDeterminer_f(s[i+1]) && (!isName_f(s[i+2])))
		return "false";
	    if((i+1)<s.length && (i+2)<s.length && isDeterminer_f(s[i+1]) && isName_f(s[i+2])){
		tree = tree + "prep_phrase(preposition("+s[i]+"), noun_phrase(determiner("+s[i+1]+"),name("+s[i+2]+")))))";
		return tree;
	    }
	    if((i+1)<s.length && (i+2)<s.length && isDeterminer_m_p(s[i+1]) && (!isName_m_p(s[i+2])))
		return "false";
	    if((i+1)<s.length && (i+2)<s.length && isDeterminer_m_p(s[i+1]) && isName_m_p(s[i+2])){
		tree = tree + "prep_phrase(preposition("+s[i]+"), noun_phrase(determiner("+s[i+1]+"),name("+s[i+2]+")))))";
		return tree;
	    }
	    if((i+1)<s.length && (i+2)<s.length && isDeterminer_f_p(s[i+1]) && (!isName_f_p(s[i+2])))
		return "false";
	    if((i+1)<s.length && (i+2)<s.length && isDeterminer_f_p(s[i+1]) && isName_f_p(s[i+2])){
		tree = tree + "prep_phrase(preposition("+s[i]+"), noun_phrase(determiner("+s[i+1]+"),name("+s[i+2]+")))))";
		return tree;
	    }
	}

	if(i<s.length && isDeterminer_m(s[i])){
	    if((i+1)<s.length && isName_m(s[i+1])){
		tree = tree + "noun_phrase(determiner("+s[i]+"),name("+s[i+1]+"))))";
		return tree;
	    }
	    else
		return "false";
	}

	if(i<s.length && isDeterminer_f(s[i])){
	    if((i+1)<s.length && isName_f(s[i+1])){
		tree = tree + "noun_phrase(determiner("+s[i]+"),name("+s[i+1]+"))))";
		return tree;
	    }
	    else
		return "false";
	}

	if(i<s.length && isDeterminer_m_p(s[i])){
	    if((i+1)<s.length && isName_m_p(s[i+1])){
		tree = tree + "noun_phrase(determiner("+s[i]+"),name("+s[i+1]+"))))";
		return tree;
	    }
	    else
		return "false";
	}

	if(i<s.length && isDeterminer_f_p(s[i])){
	    if((i+1)<s.length && isName_f_p(s[i+1])){
		tree = tree + "noun_phrase(determiner("+s[i]+"),name("+s[i+1]+"))))";
		return tree;
	    }
	    else
		return "false";
	}
	return "false";
    }

    public static void main(String[] args){
	Scanner stdin = new Scanner(System.in);
	String s = stdin.nextLine();
	String[] parsed = s.split(" ");
	System.out.println(isSentence(parsed));
    }
}

File: T3/gramatica.pl

sentence(sentence(N,V)) --> noun_phrase(N), verbal_phrase(V).
sentenca(sentenca(NP,VP)) --> noun_phrase_p(NP), verbal_phrase_p(VP).

noun_phrase(noun_phrase(Determiner, Name)) --> determiner_m(Determiner), name_m(Name).
noun_phrase(noun_phrase(Determiner, Name)) --> determiner_f(Determiner), name_f(Name).

noun_phrase(noun_phrase(Contra, Name)) --> contraction_m(Contra), name_m(Name).
noun_phrase(noun_phrase(Contra, Name)) --> contraction_f(Contra), name_f(Name).

noun_phrase_p(noun_phrase_p(Determiner, Name)) --> determiner_p_m(Determiner), name_p_m(Name).
noun_phrase_p(noun_phrase_p(Determiner, Name)) --> determiner_p_f(Determiner), name_p_f(Name).

noun_phrase_p(noun_phrase_p(Contra, Name)) --> contraction_p_m(Contra), name_p_m(Name).
noun_phrase_p(noun_phrase_p(Contra, Name)) --> contraction_p_f(Contra), name_p_f(Name).

prep_phrase(prep_phrase(Preposition,Phrase)) --> prep(Preposition), noun_phrase(Phrase).
prep_phrase(prep_phrase(Preposition,Phrase)) --> prep(Preposition), noun_phrase_p(Phrase).

verbal_phrase(verbal_phrase(Verb,Rest)) --> verb(Verb), noun_phrase(Rest).
verbal_phrase(verbal_phrase(Verb,Rest)) --> verb(Verb), noun_phrase_p(Rest).
verbal_phrase(verbal_phrase(Verb,Rest)) --> verb(Verb), prep_phrase(Rest).
verbal_phrase(verbal_phrase(Verb)) --> verb(Verb).

verbal_phrase_p(verbal_phrase_p(Verb,Rest)) --> verb_p(Verb), noun_phrase(Rest).
verbal_phrase_p(verbal_phrase_p(Verb,Rest)) --> verb_p(Verb), noun_phrase_p(Rest).
verbal_phrase_p(verbal_phrase_p(Verb,Rest)) --> verb_p(Verb), prep_phrase(Rest).
verbal_phrase_p(verbal_phrase_p(Verb)) --> verb_p(Verb).


name_m(name(cacador)) --> [cacador].
name_m(name(cachorro)) --> [cachorro].
name_m(name(rio)) --> [rio].
name_m(name(mar)) --> [mar].
name_m(name(tempo)) --> [tempo].
name_m(name(sino)) --> [sino].
name_m(name(vento)) --> [vento].
name_m(name(martelo)) --> [martelo].
name_m(name(tambor)) --> [tambor].
name_m(name(rosto)) --> [rosto].

name_f(name(menina)) --> [menina].
name_f(name(noticia)) --> [noticia].
name_f(name(floresta)) --> [floresta].
name_f(name(porta)) --> [porta].
name_f(name(vida)) --> [vida].
name_f(name(mae)) --> [mae].
name_f(name(cidade)) --> [cidade].

name_p_m(name(lobos)) --> [lobos].
name_p_m(name(tambores)) --> [tambores].

name_p_f(name(meninas)) --> [meninas].
name_p_f(name(lagrimas)) --> [lagrimas].

verb(verb(corre)) --> [corre].
verb(verb(correu)) --> [correu].
verb(verb(bateu)) --> [bateu].

verb_p(verb_p(corriam)) --> [corriam].
verb_p(verb_p(bateram)) --> [bateram].
verb_p(verb_p(correram)) --> [correram].

prep(preposition(para)) --> [para].
prep(preposition(com)) --> [com].

contraction_f(contraction(na)) --> [na].
contraction_f(contraction(pela)) --> [pela].

contraction_m(contraction(no)) --> [no].
contraction_m(contracion(pelo)) --> [pelo].

contraction_p_m(contraction(nos)) --> [nos].
contraction_p_m(contraction(pelos)) --> [pelos].

contraction_p_f(contraction(nas)) --> [nas].
contraction_p_f(contraction(pelas)) --> [pelas].

determiner_m(determiner(o)) --> [o].
determiner_m(determiner('O')) --> ['O'].

determiner_f(determiner(a)) --> [a].
determiner_f(determiner('A')) --> ['A'].

determiner_p_m(determiner(os)) --> [os].
determiner_p_m(determiner('Os')) --> ['Os'].

determiner_p_f(determiner(as)) --> [as].
determiner_p_f(determiner('As')) --> ['As'].
